src/_guides/language/concurrency/index.md

```diff
@@ -0,0 +1,456 @@
+
+[`readAsStringSync()`]: https://api.dart.dev/stable/dart-io/File/readAsStringSync.html
+[`readAsString()`]: https://api.dart.dev/stable/dart-io/File/readAsString.html
+
+Why does it matter whether a method is synchronous or asynchronous?
+It matters because most apps need to do more than one thing at a time.
+For example, an app might start an HTTP request,
+but need to update its display or respond to user input
+before the HTTP request completes.
+Asynchronous code helps apps stay responsive.
+
+### The async-await syntax
+The `async` and `await` keywords provide a declarative way to define asynchronous functions and use their results.
+
+Here’s an example of some synchronous code that blocks while waiting for file I/O:
+
+```dart
+void main() {
+  // Read some data.
+  final fileData = _readFileSync();
+  final jsonData = jsonDecode(fileData);
+
+  // Use that data.
+  print('Number of JSON keys: ${jsonData.length}');
+}
+
+String _readFileSync() {
+  final file = File(filename);
+  final contents = file.readAsStringSync();
+  return contents.trim();
+}
+```
+
+Here’s similar code, but with changes (highlighted) to make it asynchronous:
+
+{% prettify dart tag=pre+code %}
+[!Future<void>!] main() [!async!] {
+  // Read some data.
+  final fileData = [!await!] _readFileAsync();
+  final jsonData = jsonDecode(fileData);
+
+  // Use that data.
+  print('Number of JSON keys: ${jsonData.length}');
+}
+
+[!Future<String>!] _readFileAsync() [!async!] {
+  final file = File(filename);
+  final contents = [!await!] file.[!readAsString()!];
+  return contents.trim();
+}
+{% endprettify %}
+
+The `main()` function uses the `await` keyword in front of `_readFileAsync()`
+to let other Dart code (such as event handlers) use the CPU
+while native code (file I/O) executes.
+Using `await` also has the effect of
+converting the `Future<String>` returned by `_readFileAsync()` into a `String`.
+As a result, the `contents` variable has the implicit type `String`.
+
+{{site.alert.note}}
+  The `await` keyword works only in functions that
+  have `async` before the function body.
+{{site.alert.end}}
+
+As the following figure shows,
+the Dart code pauses while `readAsString()` executes non-Dart code,
+in either the Dart virtual machine (VM) or the operating system (OS).
+Once `readAsString()` returns a value, Dart code execution resumes.
+
+![Flowchart-like figure showing app code executing from start to exit, waiting for native I/O in between](/guides/language/concurrency/images/basics-await.png)
+
+If you’d like to learn more about using `async`, `await`, and futures,
+visit the [asynchronous programming codelab][].
+
+[asynchronous programming codelab]: /codelabs/async-await
+
+
+## How isolates work
+
+Most modern devices have multi-core CPUs.
+To take advantage of all those cores,
+developers sometimes use shared-memory threads running concurrently.
+However, shared-state concurrency is
+[error prone](https://en.wikipedia.org/wiki/Deadlock) and
```

Comment:
I was told in a review comment of the block post that this wikipedia link is misleading. The main problem isn't deadlocks, but data races. I'm still waiting for confirmation, but I think this link might be better:
https://en.wikipedia.org/wiki/Race_condition#In_software